open_spiel/spiel_bots.cc

```diff
@@ -179,10 +200,112 @@ std::unique_ptr<Bot> MakeFixedActionPreferenceBot(
   return std::make_unique<StatefulRandomBot>(game, player_id, seed);
 }
 
+BotRegisterer::BotRegisterer(const std::string& bot_name,
+                             std::unique_ptr<BotFactory> creator) {
+  RegisterBot(bot_name, std::move(creator));
+}
+std::unique_ptr<Bot> BotRegisterer::CreateByName(
+    const std::string& bot_name, std::shared_ptr<const Game> game,
+    Player player_id, const GameParameters& params) {
+  auto iter = factories().find(bot_name);
+  if (iter == factories().end()) {
+    SpielFatalError(absl::StrCat("Unknown bot '", bot_name,
+                                 "'. Available bots are:\n",
+                                 absl::StrJoin(RegisteredBots(), "\n")));
+
+  } else {
+    const std::unique_ptr<BotFactory>& factory = iter->second;
+    return factory->Create(std::move(game), player_id, params);
+  }
+}
+std::vector<std::string> BotRegisterer::CanPlayGame(
+    const Game& game, Player player_id, const GameParameters& params) {
+  std::vector<std::string> bot_names;
+  for (const auto& key_val : factories()) {
+    if (key_val.second->CanPlayGame(game, player_id, params)) {
+      bot_names.push_back(key_val.first);
+    }
+  }
+  return bot_names;
+}
+void BotRegisterer::RegisterBot(const std::string& bot_name,
+                                std::unique_ptr<BotFactory> creator) {
+  factories()[bot_name] = std::move(creator);
+}
+std::vector<std::string> BotRegisterer::RegisteredBots() {
+  std::vector<std::string> names;
+  for (const auto& key_val : factories()) names.push_back(key_val.first);
+  return names;
+}
+std::vector<std::string> RegisteredBots() {
+  return BotRegisterer::RegisteredBots();
+}
+bool BotRegisterer::IsBotRegistered(const std::string& bot_name) {
+  return factories().find(bot_name) != factories().end();
+}
+bool IsBotRegistered(const std::string& bot_name) {
+  return BotRegisterer::IsBotRegistered(bot_name);
+}
+std::unique_ptr<Bot> LoadBot(const std::string& bot_name,
+                             const std::shared_ptr<const Game>& game,
+                             Player player_id) {
+  return LoadBot(bot_name, game, player_id, GameParametersFromString(bot_name));
+}
+std::unique_ptr<Bot> LoadBot(const std::string& bot_name,
+                             const std::shared_ptr<const Game>& game,
+                             Player player_id, const GameParameters& params) {
+  std::unique_ptr<Bot> result =
+      BotRegisterer::CreateByName(bot_name, game, player_id, params);
+  if (result == nullptr) {
+    SpielFatalError(absl::StrCat("Unable to create bot: ", bot_name));
+  }
+  return result;
+}
+std::vector<std::string> BotsThatCanPlayGame(const Game& game,
+                                             Player player_id) {
+  return BotRegisterer::CanPlayGame(game, player_id, {});
+}
+std::vector<std::string> BotsThatCanPlayGame(const Game& game, Player player_id,
+                                             const GameParameters& params) {
```

Comment:
As noted in another comment I think it's preferable that a bot's ability to play a game is independent of the specific bot parameters. This would also solve the ambiguity of the bot parameters in this method, because I think the specific  parameters supported depend on the type of bot, so it's not clear how this can work in general across bots.

Reply:
Yes, I agree. I mentioned the original motivation in the other comment.